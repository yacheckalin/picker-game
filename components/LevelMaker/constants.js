export const MIN_MAP_SIZE = 3;
export const MAX_MAP_SIZE = 50;
export const DEFAULT_GRID_MAP = [
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
];

export const DEFAULT_CELL_SIZE = 40;
export const MARK_SELECTED = 999;
