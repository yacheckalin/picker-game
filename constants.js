export const KEYS = {
  BLUE_KEY: [20, "BLUE"],
  GREEN_KEY: [21, "GREEN"],
  RED_KEY: [22, "RED"],
};
export const DOORS = {
  BLUE_DOOR: [90, "BLUE"],
  GREEN_DOOR: [91, "GREEN"],
  RED_DOOR: [92, "RED"],
};
export const WALL = 1;
export const WALL_D = 2; // walls behind the door
export const VISITED = 8;
export const EMPTY = 0;
export const EMPTY_BACKPACK_CELL = 0;
export const DEFAULT_FIELD_SIZE = 4;
export const DEFAUTL_FIELD_HASH = "level_default";

export const MIN_MAP_SIZE = 3;
export const MAX_MAP_SIZE = 50;
export const DEFAULT_GRID_MAP = [
  [0, 0, 0],
  [0, 0, 0],
  [0, 0, 0],
];

export const DEFAULT_CELL_SIZE = 40;
export const MARK_SELECTED = 999;
export const MARK_CLEAR = 0;
