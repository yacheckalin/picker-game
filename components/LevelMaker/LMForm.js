import React, { useState } from "react";
import PropTypes from "prop-types";

import { MIN_MAP_SIZE, MAX_MAP_SIZE } from "./constants";
import { validateLevelForExport, useLocalStorage } from "../../helpers";

const LMForm = ({ generate, gridData }) => {
  const [mapSize, setMapSize] = useState(MIN_MAP_SIZE);
  const [mapHash, setMapHash] = useState("");
  const [mapMission, setMapMission] = useState("");
  const [swap, setSwap] = useState(false);
  const [isSaved, setIsSaved] = useState(false);
  const [info, setInfo] = useState(`You've got unsaved data!`);

  const [level, setLevel] = useLocalStorage(`level_${mapHash}`, ``, true);

  const handleSaveData = () => {
    let newLevel = [];
    newLevel.push(mapHash);
    newLevel.push(gridData);
    newLevel.push(parseInt(mapSize));
    newLevel.push(mapMission);

    try {
      validateLevelForExport(newLevel);
      setSwap(true);
      setLevel(JSON.stringify(newLevel));
      setInfo(`Data were copyied to localStorage!`);
      setIsSaved(true);
    } catch (e) {
      setInfo(e.message);
    }
  };

  return (
    <div className="card-panel">
      <div className="row">
        {!swap && (
          <>
            <div className="input-field col s12">
              <input
                value={mapSize}
                id="level-maker-grid-size"
                type="number"
                className="validate"
                min="3"
                max="50"
                onChange={(e) => setMapSize(e.target.value)}
              />
              <label htmlFor="level-maker-grid-size">
                Choose size {`[${MIN_MAP_SIZE} - ${MAX_MAP_SIZE}]`}:
              </label>
            </div>
            <div className="input-field col s6">
              <input
                value={mapHash}
                type="text"
                onChange={(e) => setMapHash(e.target.value)}
                id="level-maker-grid-hash"
              />
              <label htmlFor="level-maker-grid-hash">Level Hash</label>
            </div>
            <div className="input-field col s6">
              <textarea
                value={mapMission}
                type="text"
                onChange={(e) => setMapMission(e.target.value)}
                id="level-maker-grid-mission"
                className="materialize-textarea"
              />
              <label htmlFor="level-maker-grid-mission">Level Mission</label>
            </div>
          </>
        )}
      </div>

      <div className="row">
        {!swap && (
          <a
            className={`btn col s2 ${mapSize > MAX_MAP_SIZE ? `disabled` : ``}`}
            onClick={(e) => {
              generate(parseInt(mapSize));
              setIsSaved(false);
              setSwap(!swap);
            }}
          >
            Generate
          </a>
        )}
        <a className="btn col s2" onClick={(e) => setSwap(!swap)}>
          {swap ? `Show` : `Hide`}
        </a>
        {!isSaved && (
          <>
            <a className="btn" onClick={handleSaveData}>
              Save
            </a>
          </>
        )}
        <LMFormInfo message={info} />
      </div>
    </div>
  );
};

const LMFormInfo = ({ message }) => (
  <span className="secondary-content orange-text">{message}</span>
);

LMForm.propTypes = {
  generate: PropTypes.func.isRequired,
};

export default LMForm;
