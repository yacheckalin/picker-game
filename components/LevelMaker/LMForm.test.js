import React from "react";

import { shallow, mount, render, unmount } from "enzyme";
import toJSON from "enzyme-to-json";
import LMForm from "./LMForm";

import { DEFAULT_GRID_MAP } from "../../constants";

describe("<LMForm />", () => {
  const props = {
    generate: jest.fn(),
    gridData: DEFAULT_GRID_MAP,
  };
  test("renders and displays properly", () => {
    const wrapper = shallow(<LMForm {...props} />);
    expect(toJSON(wrapper)).toMatchSnapshot();
  });

  describe("validate field", () => {
    test("grid size", () => {
      const wrapper = shallow(<LMForm {...props} />);

      expect(wrapper.find("#level-maker-grid-size")).toHaveLength(1);
    });
    test("grid hash", () => {
      const wrapper = shallow(<LMForm {...props} />);

      expect(wrapper.find("#level-maker-grid-hash")).toHaveLength(1);
    });
    test("grid mission", () => {
      const wrapper = shallow(<LMForm {...props} />);

      expect(wrapper.find("#level-maker-grid-mission")).toHaveLength(1);
    });
    test("three button on a screen", () => {
      const wrapper = shallow(<LMForm {...props} />);

      expect(wrapper.find("a.btn")).toHaveLength(3);
      expect(wrapper.find("a.btn").at(0).text()).toBe("Generate");
      expect(wrapper.find("a.btn").at(1).text()).toBe("Hide");
      expect(wrapper.find("a.btn").at(2).text()).toBe("Save");
    });

    test("disable Generate/Save button when size greater then MAX_MAP_SIZE", () => {
      const wrapper = mount(<LMForm {...props} />);
      wrapper
        .find("#level-maker-grid-size")
        .simulate("change", { target: { value: "200" } });
      expect(wrapper.find("a.btn").at(0).hasClass("disabled")).toBeTruthy();
    });

    test("disable Generate/Save button when size less then MIN_MAP_SIZE", () => {
      const wrapper = mount(<LMForm {...props} />);
      wrapper
        .find("#level-maker-grid-size")
        .simulate("change", { target: { value: "2" } });
      expect(wrapper.find("a.btn").at(0).hasClass("disabled")).toBeTruthy();
      expect(wrapper.find("a.btn").at(2).hasClass("disabled")).toBeTruthy();
    });

    test("disable Generate/Save button when hash is empty", () => {
      const wrapper = mount(<LMForm {...props} />);
      wrapper
        .find("#level-maker-grid-hash")
        .simulate("change", { target: { value: "" } });
      expect(wrapper.find("a.btn").at(0).hasClass("disabled")).toBeTruthy();
      expect(wrapper.find("a.btn").at(2).hasClass("disabled")).toBeTruthy();
    });

    test("disable Generate/Save button when mission is empty", () => {
      const wrapper = mount(<LMForm {...props} />);
      wrapper
        .find("#level-maker-grid-mission")
        .simulate("change", { target: { value: "" } });
      expect(wrapper.find("a.btn").at(0).hasClass("disabled")).toBeTruthy();
      expect(wrapper.find("a.btn").at(2).hasClass("disabled")).toBeTruthy();
    });

    test("hide/show form ", () => {
      const wrapper = mount(<LMForm {...props} />);

      wrapper
        .find("#level-maker-grid-mission")
        .simulate("change", { target: { value: "mission" } });
      wrapper
        .find("#level-maker-grid-hash")
        .simulate("change", { target: { value: "hash" } });
      wrapper
        .find("#level-maker-grid-size")
        .simulate("change", { target: { value: "20" } });

      wrapper.find("a.btn").at(1).simulate("click");
      expect(wrapper.find("#level-maker-grid-mission")).toHaveLength(0);
      expect(wrapper.find("a.btn").at(0).text()).toBe("Show");

      wrapper.find("a.btn").at(0).simulate("click");
      expect(wrapper.find("#level-maker-grid-mission")).toHaveLength(1);
    });

    test("generate shows the message on a screen", () => {
      const wrapper = mount(<LMForm {...props} />);

      wrapper.find("a.btn").at(0).simulate("click");
      expect(wrapper.find("span.secondary-content").text()).toBe(
        "Map has been generated"
      );
      wrapper.unmount();
    });
  });
});
