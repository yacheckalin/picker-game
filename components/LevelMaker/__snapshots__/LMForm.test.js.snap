// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LMForm /> renders and displays properly 1`] = `
<div
  className="card-panel"
>
  <div
    className="row"
  >
    <div
      className="input-field col s12"
    >
      <input
        className="validate"
        id="level-maker-grid-size"
        max="50"
        min="3"
        onChange={[Function]}
        type="number"
        value={3}
      />
      <label
        htmlFor="level-maker-grid-size"
      >
        Choose size 
        [3 - 50]
        :
      </label>
    </div>
    <div
      className="input-field col s6"
    >
      <input
        className="validate"
        id="level-maker-grid-hash"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="level-maker-grid-hash"
      >
        Level Hash
      </label>
    </div>
    <div
      className="input-field col s6"
    >
      <textarea
        className="materialize-textarea validate"
        id="level-maker-grid-mission"
        onChange={[Function]}
        type="text"
        value=""
      />
      <label
        htmlFor="level-maker-grid-mission"
      >
        Level Mission
      </label>
    </div>
  </div>
  <div
    className="row"
  >
    <a
      className="btn col s2 disabled"
      onClick={[Function]}
    >
      Generate
    </a>
    <a
      className="btn col s2"
      onClick={[Function]}
    >
      Hide
    </a>
    <a
      className="btn"
      onClick={[Function]}
    >
      Save
    </a>
    <LMFormInfo
      message=""
    />
  </div>
</div>
`;
